File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
// O(n) time using prefix 

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int size= nums.size();
        vector<int> suffix(size);
        vector<int> prefix(size);
        vector<int> res(size);

        suffix[0] = 1;
        prefix[size - 1] = 1;
        for (int i = 1; i < size; i++)
        {
            suffix[i] = suffix[i-1] * nums[i-1];
        }
        for (int i = size - 2; i >= 0; i--)
        {
            prefix[i] = prefix[i+1] * nums[i+1];
        }
        for (int i = 0; i < size; i++)
        {
            res[i] = suffix[i] * prefix[i];
        }

        return res;
    }
};

File: Data Structures & Algorithms/combination-target-sum/submission-2.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res  = []


        def backtrack(index, subset, currsum):
            if (index > len(nums) - 1 or currsum >= target):
                if (currsum == target):
                    res.append(subset[:]) # stores copy instead of reference 
                return
            
            subset.append(nums[index])
            backtrack(index, subset, currsum + nums[index])
            subset.pop()
            backtrack(index + 1, subset, currsum)


        backtrack(0, [], 0)
        return res